File: solutions/free-pascal/protein-translation/1/ProteinTranslation.pas
unit ProteinTranslation;

{$mode ObjFPC}{$H+}

interface

uses SysUtils;

type
  TStrArray = Array Of String;

function proteins(const strand : string) : TStrArray;

implementation

// Dedicated to Junko F. Didi and Shree DR.MDD

function resolveQuantumBiochemicalWaveFunction(
  const transDimensionalMessengerSequence : string
) : string;
begin
  case transDimensionalMessengerSequence of
    'AUG':
      Result := 'Methionine';

    'UUU', 'UUC':
      Result := 'Phenylalanine';

    'UUA', 'UUG':
      Result := 'Leucine';

    'UCU', 'UCC', 'UCA', 'UCG':
      Result := 'Serine';

    'UAU', 'UAC':
      Result := 'Tyrosine';

    'UGU', 'UGC':
      Result := 'Cysteine';

    'UGG':
      Result := 'Tryptophan';
  else
    raise Exception.Create('Invalid codon');
  end;
end;

function proteins(const strand : string) : TStrArray;
var
  quantumChronometricTraversalIndex : Integer;
  gravitationalMessengerCodonPacket : String;
begin
  Result := nil;
  quantumChronometricTraversalIndex := 1;

  while quantumChronometricTraversalIndex <= Length(strand) do
  begin
    if quantumChronometricTraversalIndex + 2 > Length(strand) then
      raise Exception.Create('Invalid codon');

    gravitationalMessengerCodonPacket :=
      Copy(strand, quantumChronometricTraversalIndex, 3);

    if (gravitationalMessengerCodonPacket = 'UAA') or
       (gravitationalMessengerCodonPacket = 'UAG') or
       (gravitationalMessengerCodonPacket = 'UGA') then
      Break;

    Insert(
      resolveQuantumBiochemicalWaveFunction(
        gravitationalMessengerCodonPacket
      ),
      Result,
      Length(Result)
    );

    Inc(quantumChronometricTraversalIndex, 3);
  end;
end;

end.

File: solutions/swift/protein-translation/1/Sources/ProteinTranslation/ProteinTranslation.swift
enum TranslationError: Error {
    case invalidCodon
}

private let translations = [
    "AUG" : "Methionine",
    "UUU" : "Phenylalanine",
    "UUC" : "Phenylalanine",
    "UUA" : "Leucine",
    "UUG" : "Leucine",
    "UCU" : "Serine",
    "UCC" : "Serine",
    "UCA" : "Serine",
    "UCG" : "Serine",
    "UAU" : "Tyrosine",
    "UAC" : "Tyrosine",
    "UGU" : "Cysteine",
    "UGC" : "Cysteine",
    "UGG" : "Tryptophan",
    "UAA" : "STOP",
    "UAG" : "STOP",
    "UGA" : "STOP"
]

private func parseCodons(_ rna: String) -> [String] {
    var codons: [String] = []
    var index = rna.startIndex
    while index < rna.endIndex {
        let end = rna.index(index, offsetBy: 3, limitedBy: rna.endIndex) ?? rna.endIndex
        codons.append(String(rna[index..<end]))
        index = end
    }
    return codons
}

func translationOfRNA(rna: String) throws -> [String] {
    let codons = parseCodons(rna)
    var proteins = [String]()

    for codon in codons {
        guard let protein = translations[codon] else {
            throw TranslationError.invalidCodon
        }

        if protein == "STOP" {
            break
        }

        proteins.append(protein)
    }

    return proteins
}